From 9b94ede7e0860d3d2ad82f60d9fec46d565a4c6a Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Fri, 25 Mar 2022 05:37:56 -0400
Subject: [PATCH] MIUI13 devices hide their vibrator HAL behind non-default
 name: "vibratorfeature"

[los23-port] rebased: HIDL vibrator path removed upstream in A16; AIDL-only connectHal now uses AServiceManager, added vibratorfeature instance probe after default probe

Change-Id: I343f737a070c17d748d1ebfd34e91f7062adfef0
---
 services/vibratorservice/VibratorHalController.cpp | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/services/vibratorservice/VibratorHalController.cpp b/services/vibratorservice/VibratorHalController.cpp
index 302e3e1299..5ad4aa8729 100644
--- a/services/vibratorservice/VibratorHalController.cpp
+++ b/services/vibratorservice/VibratorHalController.cpp
@@ -56,6 +56,16 @@ std::shared_ptr<HalWrapper> connectHal(std::shared_ptr<CallbackScheduler> schedu
         }
     }
 
+    serviceName = std::string(IVibrator::descriptor) + "/vibratorfeature";
+    if (AServiceManager_isDeclared(serviceName.c_str())) {
+        std::shared_ptr<IVibrator> hal = IVibrator::fromBinder(
+                ndk::SpAIBinder(AServiceManager_waitForService(serviceName.c_str())));
+        if (hal) {
+            ALOGV("Successfully connected to Xiaomi Vibrator HAL AIDL service.");
+            return std::make_shared<AidlHalWrapper>(std::move(scheduler), std::move(hal));
+        }
+    }
+
     ALOGV("Vibrator HAL service not available.");
     gHalExists = false;
     return nullptr;
-- 
2.43.0

